fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,646 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        chunks: Number of micor-batches in a mini-batch
+        checkpoint_stop:: Checkpointing is done only for the first checkpoint_stop chunks of a mini-batch.
+    """
 
     def __init__(
         self,
-        module: List[LayerSpec],
-        *,
-        balance: List[int],
-        devices: List[str],
-        chunks: int = chunks,
-        checkpoint: str = checkpoint,
-        deferred_batch_norm: bool = False,
+        module_rref: rpc.RRef,
+        device: str,
+        num_inputs: int,
+        num_outputs: Optional[int],
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
     ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_inputs = num_inputs
+        self.num_outputs = num_outputs
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_pipeline_record(self, consumers: List[OutputConsumer]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(
+            self.device, self.rank, self.chunks, self.num_inputs, self.num_outputs, consumers
+        )
+
+    def run(self, pipeline_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism. It modifies the given batches in place."""
+        m = len(pipeline_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for chunk in range(m):
+            with record_function("feed"):
+                pipeline_record.wait_for(chunk)
+            self.fence(pipeline_record, chunk)
+            self.compute(pipeline_record, chunk)
+            # Forward outputs of processing the chunk in this parition for processing by next partition.
+            with use_stream(self.stream):
+                for consumer, input_idx, output_idx in pipeline_record.consumers:
+                    v = pipeline_record.get_batch(chunk).value[output_idx]
+                    pipeline_record.forwarded_phony[chunk][output_idx].append(
+                        consumer.remote().feed(chunk, input_idx, v)
+                    )
+
+    def fence(self, pipeline_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Prepares micro-batches for computation."""
+        # Ensure that batches[chunk-1] is executed after batches[chunk] in
+        # backpropagation by an explicit dependency.
+        # TODO: This dependency injection causes deadlock if this partition
+        # gets its input from model input. 1) Figure out why 2) If we need to live
+        # with this constraint, replace the condition 'pipeline_record.rank > 0' below with
```

Comment:
I believe the condition below is correct and matches the single-process code https://github.com/facebookresearch/fairscale/blob/cae9b638d6494fe916d0bdd308215b45473157f0/fairscale/nn/pipe/pipeline.py#L144

You do need a dependency for the first chunk or for the first partition since there is no queueing in either of these cases. The first partition becomes a sync in the backward pass and hence cannot queue.

Reply:
IIUC this dependency is to prevent queuing grads coming from second partition to the first partition. So if backward path in the first partition is slow, it is possible to we get OOM. To be honest I also had problem understanding the condition in single-process code.
But even if that's the case, it resolves "comment 1" in my comment. The second item in the comment is still valid. the condition here is not simply "rank != 0". Since with the graph some of other ranks could also have the same situation (if their input comes from model input).